### ISSUE TITLE
HTTP3 w/ HTTPS record and fallback behavior

### ISSUE THREAD

**Issue by Jxck** [h2o/h2o#3131]
I deployed HTTP3 with HTTPS record and found a fallback problem.

## tl;dr;

fallback to H2 when H3 doesn't enabled on that domain.

## details

Imagine you deploy two domains on single h2o instance.

### Before

- https://example.com (h2)
  - A
  - AAAA
- https://cdn.example.com (h2)
  - CNAME to example.com
- https://intra.example.com (h2)
  - CNAME to example.com

And h2o.conf is something like below.

```config
"example.com"
    listen:
      port: 443
      ssl:
        <<: *ssl
    listen:
      port: 80
"cnd.example.com"
    listen:
      port: 443
      ssl:
        <<: *ssl
    listen:
      port: 80
"intra.example.com"
    listen:
      port: 443
      ssl:
        <<: *ssl
    listen:
      port: 80
```

It's so standard architecture.

## After

When you wanna enable HTTP3 only for **example.com** and **cdn.example.com** only, not for **intra.example.com**.

Firstly provide HTTPS record to *example.com*.

- https://example.com (h2)
  - A
  - AAAA
  - **HTTPS**: alpn="h3"
- https://cdn.example.com (h2)
  - CNAME to example.com
- https://intra.example.com (h2)
  - CNAME to example.com

And change config.

```config
"example.com"
    listen:
      port: 443
      type: **quic**
      ssl:
        <<: *ssl
    listen:
      port: 443
      ssl:
        <<: *ssl
    listen:
      port: 80
"intra.example.com"
    listen:
      port: 443
      ssl:
        <<: *ssl
    listen:
      port: 80
"cdn.example.com"
    listen:
      port: 443
      type: **quic**
      ssl:
        <<: *ssl
    listen:
      port: 443
      ssl:
        <<: *ssl
    listen:
      port: 80
```

But this configuration doesn't works as expected.
Actually, "intra.example.com" returns "example.com" contents to the browser.

## Why

I found it happens on Chrome Beta  (107.0.5304.36).
Current Chrome Beta aware the HTTPS record with CNAME, so resolving "intra.example.com" to DNS will query HTTPS for "example.com" and considers "intra.example.com" supports H3.
And accessing "intra.example.com" with H3, but h3 doesn't supported on h2o config, so falling back to h3 supported origin from top of list. Finally, returns "example.com" contents to "intra.examplecom".

## Problem

Serve multiple domain on single server is basic architecture. Deploy DNS CNAME record for multiple domain is basic also.

In this case, enable HTTP3 one by one which is ready to serve, but HTTPS record configulation is merged in "example.com", so there are twe way to take..

1. enable http3 for all domains at the same time, then serve HTTPS => It's too tough, will suffer deploy H3.
2. serve HTTPS record for example.com and remove CNAME and serve A/AAAA etc for domain which doesn't enable H3 => CNAME is important when you serve many more multiple domain. especially it's dynamic.
3 don't serve HTTPS record for example.com and serve it for each subdomains which H3 enabled. => how to enable H3 on example.com it self ?

## Suggestion

IIUC, browser falls back to H2 when it couldn't establish quic connection to the server.
So I wondering why h2o falls back to "example.com" when "intra.example.com" doesn't support H3 ?
If h2o returns error for browsers quic handshake if specified domain (maybe in SNI) doens't H3 ready. And browser falls back to H2 to "intra.example.com" is what I expected.

Any thought, comments are welcome.
